third_party/amd/lib/TritonAMDGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -1548,14 +1548,40 @@ struct AtomicRMWOpConversion
     rmwPtr = b.ptrtoint(i64_ty, rmwPtr);
 
     auto *curBlock = rewriter.getInsertionBlock();
-    auto *afterLoopBlock = curBlock->splitBlock(rewriter.getInsertionPoint());
+    auto *leaderBlock = curBlock->splitBlock(rewriter.getInsertionPoint());
+    leaderBlock->addArgument(i64_ty, loc);
+    leaderBlock->addArgument(operandElemType, loc);
+    auto *beforeLoop = rewriter.createBlock(
+        curBlock->getParent(), std::next(Region::iterator(curBlock)));
+
+    rewriter.setInsertionPointToEnd(curBlock);
+
+    // check how many adjacent address are in the wave
+    Value rightNeighbourAddr =
+        genI32TiledOp(rewriter, shiftLeftI32ByDpp, rmwPtr);
+    Value elemSize = b.i64_val(operandElemType.getIntOrFloatBitWidth() / 8);
+    Value isNeighbour = b.icmp_eq(rightNeighbourAddr, b.add(rmwPtr, elemSize));
+    Value neighbourFlag = targetInfo.ballot(rewriter, loc, i64_ty, isNeighbour);
+    Value numNeighbours =
+        b.trunc(i32_ty, generatePopcount64(rewriter, neighbourFlag));
+    // heuristic, do optimization only if # of neighbours is less than 32,
+    // [TODO], will calculate actual # of difference addresses
```

Comment:
Nit: Please make this proper sentences and explain why such heuristic (you have that in the commit message actually).

Reply:
Changed accordingly.